File: Kap13/JavaScript/Listing39_47/mygraphs/Dreieck.mjs
import ZweiDElm from "./ZweiDElm.mjs";

export default class Dreieck extends ZweiDElm {
    #seiteA = 1;
    #seiteB = 1;
    #seiteC = 1;

    constructor(initRandfarbe, initFuellfarbe, a, b, c) {
      super(initRandfarbe, initFuellfarbe);
      console.log("Start Konstruktor Dreieck");
      if (this.checkValues(a, b, c)) {
        this.#seiteA = a;
        this.#seiteB = b;
        this.#seiteC = c;
        this.berechneFlaeche();
      }
      console.log("Ende Konstruktor Dreieck");    
    }
    
    checkValues(a, b, c) {
      if (a + b <= c || b + c <= a || c + a <= b) {
        return false;
      }
      return true;
    }

    berechneUmfang() {
        this.umfang = this.#seiteA + this.#seiteB + this.#seiteC;
    }

    berechneFlaeche() {
        this.berechneUmfang();
        const s = this.umfang/2;
        this.flaeche = Math.sqrt(s * (s - this.#seiteA) * (s - this.#seiteB) * (s - this.#seiteC));
    }

    getSeiteA() {
      return this.#seiteA;
    }
    
    setSeiteA(a) {
      if (checkValues(a, this.#seiteB, this.#seiteC)) {
        this.#seiteA = a;
        this.berechneFlaeche();
      }
    }
    getSeiteB() {
      return this.#seiteB;
    }
    
    setSeiteB(b) {
      if (checkValues(this.#seiteA, b, this.#seiteC)) {
        this.#seiteB = b;
        this.berechneFlaeche();
      }
    }
    getSeiteC() {
      return this.#seiteA;
    }
    
    setSeiteC(c) {
      if (checkValues(this.#seiteA, this.#seiteB, c)) {
        this.#seiteC = c;
        this.berechneFlaeche();
      }
    }
    
    printInfo() {
      super.printInfo();
      console.log("Seite A: " + this.#seiteA);
      console.log("Seite B: " + this.#seiteB);
      console.log("Seite C: " + this.#seiteC);
   }
  
}

File: Kap13/C#/Listing39_47/mygraphs/Dreieck.cs
namespace mygraphs {
  public sealed class Dreieck : ZweiDElm {
    private double seiteA = 1;
    private double seiteB = 1;
    private double seiteC = 1;

    public Dreieck(int initRandfarbe, int initFuellfarbe, double a, double b, double c) : base(initRandfarbe, initFuellfarbe) {
      Console.WriteLine("Start Konstruktor Dreieck");
      if (checkValues(a, b, c)) {
        seiteA = a;
        seiteB = b;
        seiteC = c;
        berechneFlaeche();
      }
      Console.WriteLine("Ende Konstruktor Dreieck");    
    }
    
    private bool checkValues(double a, double b, double c) {
      if (a + b <= c || b + c <= a || c + a <= b) {
        return false;
      }
      return true;
    }

    protected override void berechneUmfang() {
        umfang = seiteA + seiteB + seiteC;
    }

    protected override void berechneFlaeche() {
        berechneUmfang();
        double s = umfang/2;
        flaeche = Math.Sqrt(s * (s - seiteA) * (s - seiteB) * (s - seiteC));
    }

    public double getSeiteA() {
      return seiteA;
    }
    
    public void setSeiteA(double a) {
      if (checkValues(a, seiteB, seiteC)) {
        seiteA = a;
        berechneFlaeche();
      }
    }
    public double getSeiteB() {
      return seiteB;
    }
    
    public void setSeiteB(double b) {
      if (checkValues(seiteA, b, seiteC)) {
        seiteB = b;
        berechneFlaeche();
      }
    }
    public double getSeiteC() {
      return seiteC;
    }
    
    public void setSeiteC(double c) {
      if (checkValues(seiteA, seiteB, c)) {
        seiteC = c;
        berechneFlaeche();
      }
    }
    
    public override void printInfo() {
      base.printInfo();
      Console.WriteLine("Seite A: " + seiteA);
      Console.WriteLine("Seite B: " + seiteB);
      Console.WriteLine("Seite C: " + seiteC);
   }
  }	
}